lib/class-wp-block-list.php

```diff
@@ -0,0 +1,171 @@
+<?php
+/**
+ * Blocks API: WP_Block_List class
+ *
+ * @package Gutenberg
+ */
+
+/**
+ * Class representing a list of block instances.
+ */
+class WP_Block_List implements Iterator, ArrayAccess {
+
+	/**
+	 * Original array of parsed block data.
+	 *
+	 * @var array|WP_Block[]
+	 * @access protected
+	 */
+	protected $blocks;
+
+	/**
+	 * All available context of the current hierarchy.
+	 *
+	 * @var array
+	 * @access protected
+	 */
+	protected $available_context;
+
+	/**
+	 * Block type registry to use in constructing block instances.
+	 *
+	 * @var WP_Block_Type_Registry
+	 * @access protected
+	 */
+	protected $registry;
+
+	/**
+	 * Constructor.
+	 *
+	 * Populates object properties from the provided block instance argument.
+	 *
+	 * @param array|WP_Block[]       $blocks            Array of parsed block data, or block instances.
+	 * @param array                  $available_context Optional array of ancestry context values.
+	 * @param WP_Block_Type_Registry $registry          Optional block type registry.
+	 */
+	public function __construct( $blocks, $available_context = array(), $registry = null ) {
+		$this->blocks            = $blocks;
+		$this->available_context = $available_context;
+		$this->registry          = is_null( $registry ) ?
+			WP_Block_Type_Registry::get_instance() :
+			$registry;
+	}
+
+	/*
+	 * ArrayAccess interface methods.
+	 */
+
+	/**
+	 * Returns true if a block exists by the specified block index, or false
+	 * otherwise.
+	 *
+	 * @link https://www.php.net/manual/en/arrayaccess.offsetexists.php
+	 *
+	 * @param string $index Index of block to check.
+	 *
+	 * @return bool Whether block exists.
+	 */
+	public function offsetExists( $index ) {
+		return isset( $this->blocks[ $index ] );
+	}
+
+	/**
+	 * Returns the value by the specified block index.
+	 *
+	 * @link https://www.php.net/manual/en/arrayaccess.offsetget.php
+	 *
+	 * @param string $index Index of block value to retrieve.
+	 *
+	 * @return mixed|null Block value if exists, or null.
+	 */
+	public function offsetGet( $index ) {
+		$block = $this->blocks[ $index ];
+
+		if ( isset( $block ) && is_array( $block ) ) {
+			$block = new WP_Block( $block, $this->available_context, $this->registry );
```

Comment:
Note to self: This should set into `$this->blocks` so that it doesn't have to construct each time it's accessed, only the first time.

Reply:
> Note to self: This should set into `$this->blocks` so that it doesn't have to construct each time it's accessed, only the first time.

Implemented in f0bbd32b7c.